Migrate Lexbox UI to Svelte 5

rmunn opened the pull request:
Fixes #1630.
Also closes #1177, since we'll no longer need migration examples.

Migration was quite straightforward. The only difficulty was three pages that used derived stores, and also had a `$:` block that the migration tool wanted to change to `$derived`. Since there was already a variable named `derived` (the standard derived-store function from Svelte 4) this would have caused a naming conflict. I solved that by importing `derived` under the name `derivedStore` in the affected files, and this made the migration script run cleanly.

A couple of minor post-migration cleanups were needed, but nothing near as much manual cleanup as I had originally anticipated needing. Manual testing of the site showed that everything seemed to be working as it should.

rmunn commented:
Many lint errors about things like `children?: import('svelte').Snippet;` in the `$props()` produced by the migration tool. Our eslint config forbids inline imports for types, wanting them all at the top of the file. Decision is to keep the eslint rule and convert all occurrences of this error.

rmunn commented:
@myieye - There are two or three broad categories of Typescript errors in this PR that I don't understand:

1. The Svelte 5 migration tool turned a lot of our passthrough `on:event` attributes into a call to `createBubbler()` from `svelte/legacy`, but now those calls are creating type errors

For example, Form.svelte was changed as follows:

```html
Old: <form method="post" on:submit|preventDefault ... >
New: <form method="post" onsubmit={preventDefault(bubble('submit'))} ... >
```

But that causes type errors in the components that use Form, such as [ProtectedForm](https://github.com/sillsdev/languageforge-lexbox/actions/runs/14790593965/job/41526822713?pr=1634#step:9:14):

```html
<Form {enhance} on:submit>
```

> "Argument of type '\"submit\"' is not assignable to parameter of type 'never'."

2. Calls to the `children` snippet are producing type errors like the following:

```
Type '(this: void, { errors }: { errors: any; }) => any' is not assignable to type 'Snippet<[unknown]>'.
  Types of parameters '__0' and 'args_0' are incompatible.
    Type 'unknown' is not assignable to type '{ errors: any; }'. [2322]
```

I don't know where to get started with this one.

rmunn commented:
Decision made in meeting: "intermediate" components like ProtectedForm (which is a thin wrapper around Form) or the various FooButton components will acquire a rest prop so they can forward all props to the underlying component. This will involve quite a bit of work so I've bumped the size estimate up from Large to XL.

rmunn commented:
There's an infinite effect loop on the `localhost/register` page that's probably responsible for most of the Playwright test failures: I see a lot of failures where the `tempUser` fixture (which uses the register page) doesn't complete successfully. I'm tracking down the infinite loop now.

rmunn commented:
The infinite loop is in this line in PasswordStrengthMeter:

```js
run(() => { score = strength.score; }
```

If I change it to `score = untrack(() => strength.score)` then the infinite loop goes away, but the password strength meter then doesn't update. (Because `strength` is defined as `$derived(zxcvbn(password))`).

The reason we're assigning to `score` here is because we've set it to be bindable. I might have to simply change the API of PasswordStrengthMeter to take an `onScoreChanged` function prop instead, which would then be used instead of a binding. That should, I think, break the infinite effect loop we've created.

At any rate, all the Playwright tests are passing now... but the PasswordStrengthMeter component is currently broken, so that shows a hole in our test coverage.

rmunn commented:
Opened #1646 to track the fact that the Playwright tests all turned green once I broke PasswordStrengthMeter.

rmunn commented:
I created https://svelte.dev/playground/b98885d82ab5424dba49281fcd66290c?version=5.28.2 to try to reproduce the infinite loop in a minimal example, but it's not causing an infinite loop when I do it. Some other part of the code is participating in the loop besides the PasswordStrengthMeter component. Perhaps the `$tainted` store is involved somehow? More investigation needed.

myieye reviewed (commented):
Made it through a chunk. But certainly not everything.

The props are somewhat verbose in some cases. E.g.
saying optional props props can be `undefined` or assigning optional props the default value of `undefined` are both redundant. In Svelte 4 we had to explicitly assign a value (e.g. `undefined`) to make a prop truly optional, but I don't think we need to do that in Svelte 5.

~Maybe a follow up PR would be worth it to~ EDIT: I created https://github.com/sillsdev/languageforge-lexbox/issues/1651 for the following ideas:
1) Remove some of that redundancy
2) Remove usages of `createEventDispatcher`
3) Remove all usages of `svelte/legacy`. This will involve considering and preventing all the recursive update warnings:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/1pQ8O9d6a9sd0SLN9d9U/94b000be-2cec-463f-9282-01bcd797d768.png)

myieye reviewed (changes requested):
I've created https://github.com/sillsdev/languageforge-lexbox/issues/1651 to collect follow-up tasks. I tried to limit my suggestions on this PR to smaller touch-ups, but you may think some of my ideas would be better deferred to https://github.com/sillsdev/languageforge-lexbox/issues/1651.

myieye commented on frontend/src/lib/components/IconButton.svelte:
```diff
@@ -2,32 +2,53 @@
 
   const xlIcons: IconString[] = ['i-mdi-refresh'];
-  $: textSize = xlIcons.includes(icon) ? 'text-xl' : 'text-lg';
+  let textSize = $derived(xlIcons.includes(icon) ? 'text-xl' : 'text-lg');
```
I noted this in https://github.com/sillsdev/languageforge-lexbox/issues/1651

rmunn commented on frontend/src/lib/forms/Form.svelte:
```diff
@@ -1,22 +1,38 @@
+  {id}
+  use:enhanceIfRequested
+  method="post"
+  onsubmit={preventDefault(bubble('submit'))}
```
Done in commit fffcfc82.

rmunn commented on frontend/src/lib/forms/Input.svelte:
```diff
@@ -4,34 +4,66 @@
-  export let readonly = false;
-  export let error: string | string[] | undefined = undefined;
-  export let placeholder = '';
   // Despite the compatibility table, 'new-password' seems to work well in Chrome, Edge & Firefox
```
Done in commit fffcfc82.

rmunn commented on frontend/src/lib/forms/MaybeProtectedForm.svelte:
```diff
@@ -1,19 +1,35 @@
-  <slot />
-</ProtectedForm>
+  <!-- TODO: This used to have an on:submit attribute to bubble up the submit event from HTML.
+       Let's check if the createBubbler() call in Form makes that unnecessary now. -->
```
Kevin suggested adding the `...rest` to intermediate components so that callers can hook onto events bubbled up from the root components (Form in this case). But if we're removing the bubble('submit') from Form then we don't really need the `...rest` anymore.

Done in commit fffcfc82.

rmunn commented on frontend/src/lib/forms/OrgRoleSelect.svelte:
```diff
@@ -3,12 +3,18 @@
-<Select {id} bind:value label={$t('org_role.label')} {error} on:change>
+<!-- TODO: This used to have an on:change attribute to bubble up the submit event from HTML.
+     Let's check if the createBubbler() call in Form makes that unnecessary now. -->
+<Select {id} bind:value label={$t('org_role.label')} {error} {...rest}>
```
Here I think I want to keep the `...rest` in case we end up wanting to add `helplink` or `autofocus` props, or anything else defined in the lowest-level component.

Partly done in commit fffcfc82. Removed the comment, but kept the `...rest` for now.

rmunn commented on frontend/src/lib/forms/PlainInput.svelte:
```diff
@@ -7,67 +9,89 @@
-  export let readonly = false;
-  export let error: string | string[] | undefined = undefined;
-  export let placeholder = '';
   // Despite the compatibility table, 'new-password' seems to work well in Chrome, Edge & Firefox
```
Done in commit fffcfc82.

rmunn commented on frontend/src/lib/forms/ProtectedForm.svelte:
```diff
@@ -1,27 +1,34 @@
-<Form {enhance} on:submit>
-  <slot />
+<!-- TODO: This used to have an on:submit attribute to bubble up the submit event from HTML.
+     Let's check if the createBubbler() call in Form makes that unnecessary now. -->
```
Done in commit fffcfc82, and got rid of the `...rest` props.

rmunn commented on frontend/src/lib/forms/Select.svelte:
```diff
@@ -1,20 +1,38 @@
-  <select {disabled} bind:value {id} class="select select-bordered" {autofocus} on:change>
-    <slot />
+  <!-- svelte-ignore a11y_autofocus -->
+  <select {disabled} bind:value {id} class="select select-bordered" {autofocus} onchange={bubble('change')}>
```
Done in commit fffcfc82.

rmunn commented on frontend/src/lib/forms/UserTypeSelect.svelte:
```diff
@@ -3,23 +3,29 @@
-  <Select id="type" label={$t('admin_dashboard.user_filter.user_type.label')} bind:value on:change>
+  <!-- TODO: This used to have an on:change attribute to bubble up the submit event from HTML.
+       Let's check if the createBubbler() call in Form makes that unnecessary now. -->
+  <Select id="type" label={$t('admin_dashboard.user_filter.user_type.label')} bind:value {...rest}>
```
With the intermediate FooSelect components I think I want to keep the rest props so that `helplink` or `autofocus` or other props can be passed through. Wait, I think I said that already in a previous response.

Done in commit fffcfc82, except for removing `...rest`.

rmunn commented on frontend/src/routes/(authenticated)/project/[project_code]/BulkAddProjectMembers.svelte:
```diff
@@ -89,89 +97,97 @@
-        <p class="flex gap-1 items-center">
-          <Icon icon="i-mdi-creation-outline" color="text-success" />
-          {$t('project_page.bulk_add_members.accounts_created', {createdCount: createdMembers.length})}
+        <PasswordStrengthMeter password={$form!.password} />
```
Good catch; thanks. Done in commit fffcfc82.

rmunn commented on frontend/src/lib/layout/DetailsPage.svelte:
```diff
@@ -1,30 +1,65 @@
+  {#snippet title()}
+    {@render titleRender?.()}
+  {/snippet}
+  {#snippet headerContent()}
```
Done in commit 7ed39979.

rmunn replied:
@myieye, you might want to look at this one before I resolve it, the code changes were almost trivial but not quite. But I think I got it right.

myieye replied:
I refactored it a tad in [33f977b](https://github.com/sillsdev/languageforge-lexbox/pull/1634/commits/33f977b5106ced47488ab497a3b33b362ed0cb4c):
- This change let me keep the `title` props seperate which keeps their types simpler: `export interface Props extends Omit<PageProps, 'title'>`
- Then I just removed a bunch of props that were still declared in multiple prop types.

rmunn commented:
@myieye - Addressed all review comments. I didn't resolve them because Kevin prefers to resolve comments once he's satisfied with how I addressed them, and it's easier to see the unresolved ones than to open resolved ones to check if they were addressed satisfactorily. I figured you'd prefer the same, but if you want me to resolve them as I address them then just let me know and I can do that.

myieye commented on frontend/src/lib/components/TusUpload.svelte:
```diff
@@ -10,31 +11,46 @@
+    inputDescription = undefined,
+    internalButton = false,
+  }: Props = $props();
   const dispatch = createEventDispatcher<{
```
I put it in https://github.com/sillsdev/languageforge-lexbox/issues/1651

rmunn commented on frontend/src/lib/components/Projects/ProjectConfidentialityFilterSelect.svelte:
```diff
@@ -14,10 +18,11 @@
 
 <div class="relative">
-  <Select label={$t('project.confidential.confidentiality')} helpLink={helpLinks.confidentiality} bind:value on:change>
+  <!-- TODO: Used to have an on:change attribute below, let's remove it and see if the bubbler function works as it should -->
```
Done when I went through and worked on rest props for Select components. Commit 35cd59bc, I believe.

myieye commented on frontend/src/lib/forms/Select.svelte:
```diff
@@ -4,7 +4,7 @@
   import { randomFormId } from './utils';
 
-  interface Props {
+  export interface Props {
```
For next time, I would give/export these props with a more specific name, so that importers don't have to rename it themselves.

rmunn replied:
Done in commit be1ef144d

myieye commented on frontend/src/lib/forms/UserTypeSelect.svelte:
```diff
@@ -18,7 +18,8 @@
 
 <div class="relative">
-  <Select id="type" label={$t('admin_dashboard.user_filter.user_type.label')} bind:value {...rest}>
+  <!-- Note: important that ...rest be at the start of the attributes list so that `label={$t(...)}` will not be overridden by the `label` prop in ...rest -->
```
For next time, you could also solve this with: `interface Props extends Omit<SelectProps, 'label'>`.

EDIT: actually, that's probably what you want to do to fix all the remaining eslint errors

rmunn replied:
Weird that I'm not getting any lint errors when I run `pnpm lint`. I'll go ahead and do `Omit<..., 'label'>` everywhere anyway, but I might not be able to see the lint errors get fixed until after pushing. So it's possible I'll be pushing a few commits here...

rmunn replied:
Done in commit ed306c4f7.

myieye replied:
Oh, they were svelte check errors, not lint errors.

rmunn commented on frontend/src/lib/components/DeleteUserModal.svelte:
```diff
@@ -4,12 +4,16 @@
 
-  export async function open(user: {id: string, name: string}): ReturnType<(typeof modal)['open']> {
-    return await modal.open(user.name, async () => {
+  export async function open(user: { id: string; name: string }): ReturnType<Exclude<typeof modal, undefined>['open']> {
```
Done in commit c9c93b27.

rmunn commented on frontend/src/lib/components/TusUpload.svelte:
```diff
@@ -147,12 +163,14 @@
           bind:this={fileInput}
-          on:cancel|stopPropagation
-          on:change={fileChanged}
+          oncancel={stopPropagation(bubble('cancel'))}
```
Done in commit c9c93b27.

rmunn commented on frontend/src/lib/components/Projects/WritingSystemBadge.svelte:
```diff
@@ -1,13 +1,21 @@
+  let variant: BadgeVariant = $state('badge-info');
+  run(() => {
+    variant = isActive ? (isDefault ? 'badge-primary' : 'badge-info') : 'badge-neutral';
+  });
```
Done in commit c9c93b27.

rmunn commented:
@myieye - Please see if you can reproduce the behavior I just saw where a project page loaded fine when client-side rendered (e.g. clicking from the admin dashboard) but threw a server-side error when server-side rendered (e.g. refresh browser tab when on the project page).

myieye commented on frontend/src/lib/layout/DetailsPage.svelte:
```diff
@@ -16,6 +16,8 @@
     ...rest
   }: Props = $props();
+
+  const headerContentRender = $derived(headerContent);
```
For next time: you probably could have made this change directly in the list of props with: `headerContent: headerContentRender,`. I.e. you can rename a property like that.

rmunn replied:
I'll double check, but I suspect that would have still triggered the infinite update loop, i.e. the key to breaking the loop was not to rename it but to have a different object. But if renaming it was all that was needed, then yes, that would have been simpler.
